org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/filesystem/JdtlsFsUtils.java

```diff
@@ -0,0 +1,182 @@
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Microsoft Corporation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.core.internal.filesystem;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+import org.eclipse.core.internal.preferences.EclipsePreferences;
+import org.eclipse.core.resources.IContainer;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IProjectDescription;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IJavaProject;
+
+/**
+ * Utilities of the file system implementation.
+ */
+public class JdtlsFsUtils {
+    private JdtlsFsUtils() {}
+
+    public static final IPath METADATA_FOLDER_PATH = ResourcesPlugin.getPlugin().getStateLocation().append(".projects");
+
+    /**
+     * The system property key to specify the file system mode.
+    */
+    public static final String GENERATES_METADATA_FILES_AT_PROJECT_ROOT = "java.import.generatesMetadataFilesAtProjectRoot";
+
+    public static final String FACTORY_PATH = ".factorypath";
+
+    /**
+     * The metadata files
+     */
+    private static final Set<String> METADATA_NAMES = new HashSet<>(Arrays.asList(
+        IProjectDescription.DESCRIPTION_FILE_NAME,
+        EclipsePreferences.DEFAULT_PREFERENCES_DIRNAME,
+        IJavaProject.CLASSPATH_FILE_NAME,
+        FACTORY_PATH
+    ));
+
+    /**
+     * Determine whether the resource should be stored in workspace's metadata folder.
+     * <p>
+     * The file will be stored in workspace's metadata folder when following conditions meet:
+     * <ul>
+     *   <li>The system property shows that it's allowed to store them in workspace.</li>
+     *   <li>The file belongs to the metadata file defined in {@link JdtlsFsUtils#METADATA_NAMES}.</li>
+     *   <li>The project's root path does not contain the metadata file with the same name.</li>
+     * </ul>
+     * </p>
+     * 
+     * @param location the path of the resource.
+     * @return whether the resource needs to be stored in workspace's metadata folder.
+     */
+    static boolean shouldStoreInWorkspaceStorage(IPath location) {
+        if (generatesMetadataFilesAtProjectRoot()) {
+            return false;
+        }
+
+        if (location == null || location.segmentCount() < 2) {
+            return false;
+        }
+
+        if (location.lastSegment().endsWith(EclipsePreferences.PREFS_FILE_EXTENSION)) {
+            location = location.removeLastSegments(1);
+        }
+
+        if (location.segmentCount() < 2) {
+            return false;
+        }
+
+        if (!METADATA_NAMES.contains(location.lastSegment())) {
+            return false;
+        }
+
+        if (!isAtProjectRoot(location.removeLastSegments(1))) {
+            return false;
+        }
```

Comment:
I suggest to extract L73-L91 to a new function as `isProjectMetadataFile()`, that is easier to understand.